core/charm/channel.go

```diff
@@ -0,0 +1,142 @@
+// Copyright 2020 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package charm
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/juju/errors"
+)
+
+// Risk describes the type of risk in a current channel.
+type Risk string
+
+const (
+	Stable    Risk = "stable"
+	Candidate Risk = "candidate"
+	Beta      Risk = "beta"
+	Edge      Risk = "edge"
+)
+
+var risks = []Risk{
+	Stable,
+	Candidate,
+	Beta,
+	Edge,
+}
+
+func isRisk(potential string) bool {
+	for _, risk := range risks {
+		if potential == string(risk) {
+			return true
+		}
+	}
+	return false
+}
+
+// Channel identifies and describes completely a store channel.
+type Channel struct {
+	Track  string
+	Risk   Risk
+	Branch string
+}
+
+// MakeChannel creates a core charm Channel from a set of component parts.
+func MakeChannel(track, risk, branch string) (Channel, error) {
+	if !isRisk(risk) {
+		return Channel{}, errors.NotValidf("risk %q", risk)
+	}
+	return Channel{
+		Track:  track,
+		Risk:   Risk(risk),
+		Branch: branch,
+	}, nil
+}
+
+// ParseChannel parses a string representing a store channel.
+// The returned channel's track, risk and name are normalized.
+func ParseChannel(s string) (Channel, error) {
+	if s == "" {
+		return Channel{}, errors.Errorf("channel cannot be empty")
+	}
+
+	p := strings.Split(s, "/")
+
+	var risk, track, branch *string
+	switch len(p) {
+	case 1:
+		if isRisk(p[0]) {
+			risk = &p[0]
+		} else {
+			track = &p[0]
+		}
+	case 2:
+		if isRisk(p[0]) {
+			risk, branch = &p[0], &p[1]
+		} else {
+			track, risk = &p[0], &p[1]
+		}
+	case 3:
+		track, risk, branch = &p[0], &p[1], &p[2]
+	default:
+		return Channel{}, errors.Errorf("channel is malformed and has to many components %q", s)
+	}
```

Comment:
AFAIK, you can't have a track without a risk. I'm wondering if "https://snapcraft.io/docs/channels" is up to date. It seems to indicate that you can have a risk without a track, however I've never seen that in practice. If you specify a risk without a track, it means the track is "latest". I do not know if "latest" is a snap cmd output thing.

Reply:
So you can, but it's normalised to have a track.